### PR TITLE
Payment functionality

### PR DESCRIPTION
**Pull request by ferchuhill**
Add the payment functionality. In this transaction, mark the container as completed, generate a new payment transaction, and send the payment crypto to circular API

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM node:lts
+FROM node:10.15.3
```
Comment (santiagosemhan) [inline Dockerfile:1]: Change it to node v14.

File: docker-compose.yml [hunk 2]
```diff
@@ -12,6 +12,25 @@ services:
       default:
         aliases:
           - redis-server
+  mongo:
+    image: mongo:4.2
+    restart: always
+    volumes:
+      - ${MONGO_VOLUMEN}:/data/db:cached
```
Comment (santiagosemhan) [inline docker-compose.yml:19]: change it to MONGO_VOLUME

File: src/cloud/services/WalletServices.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+interface Balance {
+  accountName: string;
+  balance: number;
+  lastSeqNum: number;
+}
+
+const getBalance = async (account: string): Promise<Balance> => {
+  const response: Balance = await Axios.get(
+    `http://api.sandbox.circularnetwork.io/v1/project/${CRYPTOS_SYMBOL_TYPES.JELLYCOIN}/users/${account}`,
```
Comment (santiagosemhan) [inline src/cloud/services/WalletServices.ts:18]: Base url: `http://api.sandbox.circularnetwork.io/v1/project/` must be on an ENV_VAR.

File: src/cloud/services/WalletServices.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    sb.pushAsset(`${formatAmount} ${CRYPTOS_SYMBOL_TYPES.JELLYCOIN}`);
+    sb.pushString(motive);
+    sb.pushNumberAsUint64(lastSeqNum);
+    sb.pushNumberAsUint64(1);
+
+    const buff = Buffer.from(sb.asUint8Array());
+    const signature = ecc.sign(buff, privkey);
+
+    await Axios.post(
+      `http://api.sandbox.circularnetwork.io/v1/project/${CRYPTOS_SYMBOL_TYPES.JELLYCOIN}/transfer`,
```
Comment (santiagosemhan) [inline src/cloud/services/WalletServices.ts:69]: same here, pull it off from the ENV_VAR

File: src/cloud/services/WorkflowService.ts [hunk 5]
```diff
@@ -829,26 +831,128 @@ const registerTransportReject = async (
   }
 };
 
+/**
+ * Check if the recicling center has the balance to pay the transaction
+ * @param containersPayment
+ * @param recyclingCenter
+ * @param user
+ */
+const checkBalanceTransaction = async ({
```
Comment (santiagosemhan) [inline src/cloud/services/WorkflowService.ts:840]: replace fn name by `throwInsufficientBalance`

File: src/cloud/services/WorkflowService.ts [hunk 6]
```diff
@@ -829,26 +831,128 @@ const registerTransportReject = async (
...
+  });
+
+  // Check the balance of the recycling center
+  const { balance } = await getBalance(recyclingCenter.get('walletId'));
+  if (balance <= totalPayment + ammountTransport) {
+    throw new Error(`The Recycling Center has not enough money, the balance is ${balance} jyc`);
+  }
+};
+
+const paymentCrypto = async ({
```
Comment (santiagosemhan) [inline src/cloud/services/WorkflowService.ts:900]: replace to `executePayment`

File: src/cloud/services/WalletServices.ts (hunk 4 above)
Comment (santiagosemhan) [inline src/cloud/services/WalletServices.ts:69]: replace here by the env var

File: src/cloud/services/WorkflowService.ts [hunk 7]
```diff
@@ -829,26 +831,128 @@ const registerTransportReject = async (
...
+      transaction.forEach((historyTransaction) => {
+        ammountTransport += historyTransaction.retribution.estimated;
+      });
+    }
+  });
+
+  // Check the balance of the recycling center
+  const { balance } = await getBalance(recyclingCenter.get('walletId'));
+  if (balance <= totalPayment + ammountTransport) {
+    throw new Error(`The Recycling Center has not enough money, the balance is ${balance} jyc`);
```
Comment (santiagosemhan) [inline src/cloud/services/WorkflowService.ts:896]: hide the balance. Show only "insufficient funds"

File: src/cloud/utils/cryptography.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+const { randomBytes, createCipheriv, createDecipheriv } = require('crypto');
+
+// const iv = Buffer.from(originalIv);
```
Comment (santiagosemhan) [inline src/cloud/utils/cryptography.ts:3]: remove comment code

**Review by santiagosemhan — CHANGES_REQUESTED** [review]
Fix comments

**Comment by ferchuhill** [comment]
all the comments are made

File: Dockerfile (hunk 1 above)
Comment (ferchuhill) [reply Dockerfile]: done

File: src/cloud/services/WalletServices.ts (hunk 3 above)
Comment (ferchuhill) [reply src/cloud/services/WalletServices.ts]: done

File: src/cloud/services/WalletServices.ts (hunk 4 above)
Comment (ferchuhill) [reply src/cloud/services/WalletServices.ts]: done

File: src/cloud/services/WorkflowService.ts (hunk 5 above)
Comment (ferchuhill) [reply src/cloud/services/WorkflowService.ts]: done

File: src/cloud/services/WorkflowService.ts (hunk 6 above)
Comment (ferchuhill) [reply src/cloud/services/WorkflowService.ts]: done

File: src/cloud/services/WalletServices.ts (hunk 4 above)
Comment (ferchuhill) [reply src/cloud/services/WalletServices.ts]: done

File: src/cloud/services/WorkflowService.ts (hunk 7 above)
Comment (ferchuhill) [reply src/cloud/services/WorkflowService.ts]: done

File: src/cloud/utils/cryptography.ts (hunk 8 above)
Comment (ferchuhill) [reply src/cloud/utils/cryptography.ts]: done

File: .env_docker.example [hunk 9]
```diff
@@ -26,5 +26,6 @@ SENDGRID_API_KEY=
 GOOGLE_MAPS_API_KEY=
 
 CIRCULAR_API_KEY=
+CIRCULAR_BASE_URL=
```
Comment (santiagosemhan) [inline .env_docker.example:29]: is this new env var on docker compose file?

**Review by santiagosemhan — APPROVED** [review]

File: .env_docker.example (hunk 9 above)
Comment (ferchuhill) [reply .env_docker.example:29]: yes

**Pull request merged**